### PR TITLE
Fix #732 - Refactor currency classes to reduce duplication

### LINKED ISSUE #732: Refactor currency types to eliminate duplication
Refactor currency types to eliminate duplication and enable direct comparisons

### PR DESCRIPTION
**Pull request by trizin**
Fixes #732

Changes proposed in this PR:

- Refactor currency classes to reduce duplication
- Enforce same unit in comparisons
- Specify types in all comparisons.
- New tests

### CODE REVIEW

**Comment by trentmc** [comment]
@calina-c I defer to you to review this, since you've been more deeply enmeshed in it

File: pdr_backend/publisher/test/test_publish_assets.py [hunk 1]
```diff
@@ -28,15 +29,15 @@ def _test_barge(network, monkeypatch):
 
     n_calls = len(mock_publish_asset.call_args_list)
     assert n_calls == 1 * 3
-
+    print(mock_publish_asset.call_args_list)
```
Comment (calina-c) [inline pdr_backend/publisher/test/test_publish_assets.py:32]: do not create print statements. Please use logging if needed.

File: pdr_backend/util/currency_types.py [hunk 2]
```diff
@@ -1,147 +1,144 @@
...
-
-    def to_wei(self) -> "Wei":
-        return Wei(int(self.amt_eth * 1e18))
+class EthUnit(ABC):
+    def __init__(self, amount: Union[int, float]):
+        self.amount = amount
 
     def __str__(self) -> str:
-        return f"{self.amt_eth} eth"
+        return f"{self.amount}"
```
Comment (calina-c) [inline pdr_backend/util/currency_types.py:15]: I would keep the quantifier to differentiate that it is not a regular number, when printing. But I'm fine either way.

File: pdr_backend/util/currency_types.py [hunk 3]
```diff
@@ -1,147 +1,144 @@
...
 
     def __repr__(self) -> str:
         return self.__str__()
 
-    def __float__(self) -> float:
-        return self.amt_eth
+    def __eq__(self, other) -> bool:
+        if isinstance(other, EthUnit) and type(self) is type(other):
+            return self.to_wei().amount == other.to_wei().amount
+        if other == 0:
```
Comment (calina-c) [inline pdr_backend/util/currency_types.py:23]: why do we need special handling for zeroes?

**Review by calina-c — CHANGES_REQUESTED** [review]

File: pdr_backend/util/currency_types.py [hunk 4]
```diff
@@ -1,147 +1,144 @@
...
 
-        return Eth(self.amt_eth / other)
 
-    def __mul__(self, other) -> "Eth":
-        if isinstance(other, Eth):
-            return Eth(self.amt_eth * other.amt_eth)
+@enforce_types
+class Eth(EthUnit):
+    def __init__(self, amt_eth: Union[int, float]):
+        super().__init__(amt_eth)
```
Comment (calina-c) [inline pdr_backend/util/currency_types.py:100]: I would reverse the order of these statements. First validate, then call super

**Pull request merged**